### DISCUSSION TITLE
Possibility of 2d simulation

### DISCUSSION THREAD

**Discussion by Hexlord** [jrouwe/JoltPhysics#180] [Q&A] [answered]
Is it possible to constrain bodies to remain fixed in one of the axis-es? This would allow using JoltPhysics for 2.5D and 2D games, which is what I wonder about doing.

**Comment by jrouwe** [comment]
Hello,

I think this is certainly possible (although I never tried), but I would only recommend this if you want to add some 3D physics 'visual fluff'. If all action takes place in 2D then you might be better off with Box2D as a simulation in 2D is a LOT simpler (and cheaper) than one in 3D.

Let's say your action takes place in the XY plane, then you would make sure that all your shapes are the same size in Z (say -1 to 1 meter) and that all shapes have the same shape along Z (so you wouldn't use a sphere but a cylinder around Z etc.). Next, when you spawn any bodies, make sure their initial position Z=0 and their rotation around X and Y is also 0. When the (dynamic) body has been constructed, you also want to override the inverse inertia matrix so that X and Y are 0 using body->GetMotionProperties()->SetInverseInertia(Vec3(0, 0, value), Quat::sIdentity()). This will prevent the objects from rotating along any other axis than Z. 'value' can be obtained by body->GetMotionProperties()->GetLocalSpaceInverseInertia().GetColumn(2).Length().

My guess is that if you do all of this, there will still be some numerical drift that slowly moves bodies away from Z=0, so you may need to update the Z positions for the bodies every now and then to put them back to 0.

**Comment by jankrassnigg** [comment]
PhysX has a feature where you can lock individual axis per body. So you can prevent a body from moving along Z and rotating around X and Y.

Sounds like a nice feature. In practice, it's pretty useless. The problem is, that it completely breaks friction somehow. Once a body rests on the ground, it's not easy to push it away anymore without an upward force. I guess that their implementation just discards values along the locked axis and thus a lot of energy gets lost. I didn't manage to get usable results out of that feature.

My point is, if you really need proper 2D physics, use Box2D. As far as I know it's a damned good library. Abusing a 3D physics engine for this may be more trouble than it's worth.

**Comment by Hexlord** [comment, marked as answer]
My reasoning was to use Jolt for both 2D and 3D games in a single engine so that I have a single physics flow (read locks, concurrent queries, etc.), but you are right, for performance reasons I should stick to Box2D for 2D physics, thank you!

**Comment by meshula** [comment]
A hack I've used to get 2d in a 3d sim is to construct the 2d objects as extrusions, and rest them on an inclined plane, much like children's cut-out toys on a board. 

It's worth exploring the many 2d engines out there, Box dates from a period with quite different memory access characteristics than today's systems (I had to fundamentally rewrite the guts to get the performance I needed on ARM). You may find a good match in Chipmunk, or even piecing one together from single-header-libraries (e.g. https://gamedevelopment.tutsplus.com/tutorials/how-to-create-a-custom-2d-physics-engine-the-basics-and-impulse-resolution--gamedev-6331).

**Comment by jrouwe** [comment]
Hello,

I created a branch for being able to lock bodies to the XY plane:

https://github.com/jrouwe/JoltPhysics/tree/feature/2d_sim

I'm not entirely happy with it (basically it loses energy so it is not the best solution and a better solution will waste CPU cycles for 3D simulations) so I'm not sure if I will merge it in this form. It is usable for experiments though.

**Reply by KTRosenberg** [reply]
I was considering using this library for 2D and found this discussion, but it looks like, as you mentioned, there are energy loss issues. Would you have any thoughts on what would be required to solve that issue? I wonder if there could be 2D equivalent functions created to avoid the extra cycles too. I think a very cool use case involves not only locking to a plane, but also locking to a spline path. You see that a lot in 2.5D games. (It's still mostly 2D in the end though when it comes to locking things to paths.)

**Reply by jrouwe** [reply]
If you want to lock things to a spline then you could use a regular 3D simulation in combination with a PathConstraint class. Objects that only exist in a 2D plane could be implemented using a SixDOFConstraint that limits the movement to two linear axis and 1 rotation axis. If you do this then energy will be conserved but as constraints are not free you will pay some extra overhead.

The reason for the energy loss on the feature/2d_sim branch is that any movement in the Z direction will simply be discarded. If your shapes are proper 2D shapes (e.g. a circle should be implemented as a cylinder and not a sphere) and if you attach all constraints to Z=0 then there shouldn't be much energy loss because (contact) constraints should not generate velocities in Z direction. To fix this properly, the mass of an object should be changed from a float to a vector so that the mass in Z direction can be made infinite, but this changes all constraint calculations and makes them more complex so will waste CPU time for the 3D case. Jolts primary use case is for 3D simulation so I'd rather not make compromises on speed there.

**Reply by KTRosenberg** [reply]
Thanks for clarifying. Are there examples showing how to use these constraint types, and if not, something similar?
Otherwise, it seems straightforward to use constraints.
I understand the preference for 3D optimization. I suppose down the line there could be specialization, but that doesn’t sound like a good idea now, agreed.

**Reply by jrouwe** [reply]
I created an example of using a constraint to limit objects to a 2d plane in #378.

**Reply by KTRosenberg** [reply]
Thanks very much!

**Reply by hhyyrylainen** [reply]
I have a problem with the approach shown in https://github.com/jrouwe/JoltPhysics/pull/378/files My physics bodies don't really want to stay at Y=0, instead they drift up and down, and are slow to return to the desired Y-coordinate. Also the rotation lock doesn't seem to work. I found this other place talking about this: https://github.com/jrouwe/JoltPhysics/issues/359#issuecomment-1384259706 from which I was able to grab a working rotation lock. So this is my working method (I removed some extra checks etc. which is not required for this to function):
```c++
JPH::BodyLockWrite lock(physicsSystem->GetBodyLockInterface(), body.GetId());
if (!lock.Succeeded())
    return nullptr;

JPH::SixDOFConstraintSettings constraintSettings;

if (axis.GetX() != 0)
    constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::TranslationX);

if (axis.GetY() != 0)
    constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::TranslationY);

if (axis.GetZ() != 0)
    constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::TranslationZ);


if (axis.GetX() != 0)
{
     constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::RotationY);
     constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::RotationZ);
}

if (axis.GetY() != 0)
{
     constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::RotationX);
     constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::RotationZ);
}

if (axis.GetZ() != 0)
{
     constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::RotationY);
     constraintSettings.MakeFixedAxis(JPH::SixDOFConstraintSettings::RotationZ);
}

// Non-working rotation lock:
/*
JPH::MassProperties mass_properties = lock.GetBody().GetShape()->GetMassProperties();
JPH::MotionProperties* mp = lock.GetBody().GetMotionProperties();
mp->SetInverseMass(1.0f / mass_properties.mMass);

// Start off with allowing all rotation
JPH::Vec3 inverseInertiaVector = JPH::Vec3(1.0f / mass_properties.mInertia.GetAxisX().Length(),
    1.0f / mass_properties.mInertia.GetAxisY().Length(), 1.0f / mass_properties.mInertia.GetAxisZ().Length());

// And then remove the axes that are not allowed
if (axis.GetX() != 0)
{
     inverseInertiaVector.SetY(0);
     inverseInertiaVector.SetZ(0);
}

if (axis.GetY() != 0)
{
     inverseInertiaVector.SetX(0);
     inverseInertiaVector.SetZ(0);
}

if (axis.GetZ() != 0)
{
     inverseInertiaVector.SetX(0);
     inverseInertiaVector.SetY(0);
}

mp->SetInverseInertia(inverseInertiaVector, JPH::Quat::sIdentity());
*/


auto constraintPtr = (JPH::SixDOFConstraint*)constraintSettings.Create(JPH::Body::sFixedToWorld, lock.GetBody());

physicsSystem->AddConstraint(constraintPtr.GetPtr());
```

For now I've increased the vertical thickness of my physics shape (to solve the problem I had: https://github.com/jrouwe/JoltPhysics/issues/568) and added manual code to check the Y-positions of bodies and if they are too far way from Y=0, forcing their position back to that. This code seems to trigger each frame for multiple bodies, so a lot of them are drifting from the desired locked axis quite a bit, which is fine when a bunch of stuff is colliding but the real problem is that the positions don't get automatically fixed by Jolt for a long time even after a collision ends that caused the pushing along the Y-axis.

Did I adapt the code wrong to fit or could it be that as I try to rotate the bodies to point certain way with angular velocity that messes with the inertia somehow? Well that's not the critical point as I can use the different approach with seemingly no problem. What I'd like to definitely know is why the Y-axis drift is happening and if it is expected or can I do something to fix it (other than manually resetting the y-position almost each frame)?

**Reply by jrouwe** [reply]
I'm not sure what the meaning of 'axis' is in your code and what values it can assume. If it is only ever (1, 0, 0), (0, 1, 0) or (0, 0, 1) then the rotation code looks ok-ish:

```
JPH::Vec3 inverseInertiaVector = JPH::Vec3(1.0f / mass_properties.mInertia.GetAxisX().Length(),
    1.0f / mass_properties.mInertia.GetAxisY().Length(), 1.0f / mass_properties.mInertia.GetAxisZ().Length());
```

Will basically make the inertia tensor a diagonal matrix, so your 'free rotation' case is already messing with the tensor.

The thing you're missing that could be causing the slow convergence is:

```
constraintSettings.mPosition1 = constraintSettings.mPosition2 = lock.GetBody().GetCenterOfMassPosition(); // Attach it at the initial position of the body for best precision
```

(take a look at the sample TwoDFunnelTest)

**Reply by hhyyrylainen** [reply]
I see the master branch had some extra code added to the funnel sample, which wasn't in that original PR code I looked at.

I added:
```
constraintSettings.mPosition1 = constraintSettings.mPosition2 = lock.GetBody().GetCenterOfMassPosition();
```
Which improved the accuracy a bit but there was still quite a lot of bouncing up and down happening.

> I'm not sure what the meaning of 'axis' is in your code and what values it can assume. If it is only ever (1, 0, 0), (0, 1, 0) or (0, 0, 1) then the rotation code looks ok-ish:

Yes, it's a vector for specifying which axes to lock, for now I'm only really interested in Y-axis lock (so the value is always (0, 1, 0)). I took that approach from https://github.com/jrouwe/JoltPhysics/issues/359 as it seemed like it would allow the potential for locking 2 axes at once.

> Will basically make the inertia tensor a diagonal matrix, so your 'free rotation' case is already messing with the tensor.

I see. For clarity I removed the part where the rotation algorithm is picked. I have a parameter to disable the locking entirely so that shouldn't matter.

I'll probably remove the inertia lock approach anyway as I can't get it to work stably at all. Even after adding that constraint settings position it is very unstable. For some reason adding `inverseInertiaVector = JPH::Vec3(0, 1.0f / mass_properties.mInertia.GetAxisY().Length(), 0);` to override the logic I had before for picking the axis for some reason results in a more stable approach, but it still causes entities to wobble a lot and rotate also on the wrong axis. The rotation settings make fixed axis seems to work much better so I'll just use that as the only option. 

So in the end I'll use the `constraintSettings.MakeFixedAxis` to fix movement axes and rotation axes. And then use external code to correct the positions of bodies that difted from the wanted Y-value. That seems stable and doesn't look like it has any problems.

**Comment by jrouwe** [comment]
Hello,

I just integrated #602 which makes it possible to have a 2D simulation by just setting a few flags on a body.

Regards,

Jorrit.

**Reply by hhyyrylainen** [reply]
Thank you for making this an official feature! I just got the chance to switch my code over to the new way to do the axis lock, and it's so much better.

I did run into some controlled rotation jittering around the target angle, but I just realized that if I zero out the locked axes in the rotation that gets entirely rid of the jitter. So if anyone with the same problem happens to read this, try only specifying the rotation around the free axis and that should make things smooth. I guess there is some kind of really small rotation value on the locked axes and when setting some angular velocity with those, it causes very apparent jittering motion when rotating for some reason.

**Comment by finscn** [comment]
@jrouwe I found the the 2d version of https://rapier.rs/  is "real 2d".
It replaces all 3D computing with 2D computing (algorithm  , vec, matrix , etc.) when build the source with 2D arguments ( by macro & some magic build-scripts ) .

like this: 

![image](https://github.com/jrouwe/JoltPhysics/assets/288367/2801287c-e2b7-4344-a73d-04d6c1d183df)


Would Jolt do that in the future ?

**Reply by hhyyrylainen** [reply]
I need both 2D and 3D physics in the same project, so I hope the already added 2D functionality is not deprecated or removed in favour of this type of approach.

**Reply by jrouwe** [reply]
It's extremely unlikely that I would do that. It would mean a major increase in code complexity. I would suggest Box2D if you want a 2D only physics engine.
